compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/extended/MultiGuardNode.java

```diff
@@ -69,6 +69,20 @@ public Node canonical(CanonicalizerTool tool) {
         } else if (guards.size() == 1) {
             // Only a single guard left => replace multi-guard with that single guard.
             return guards.get(0);
+        } else {
+            // Merge chain of MultiGuardNodes.ra
```

Comment:
`canonical` methods have the contract that they shouldn't mutate the node so this should rather return a new node.

Strictly speaking `guards.trim()` above is a mutation but i suppose it's accepted because it only changes some internal representation without being observable in meaningful ways outside this node.

Reply:
I reworked it so it returns a new node